File: backend-spring/TiendaYA/src/main/java/com/tiendaya/controllers/ProductoController.java
package com.tiendaya.controllers;

import com.tiendaya.dtos.CreateProductoDto;
import com.tiendaya.dtos.MessageDto;
import com.tiendaya.dtos.UpdateProductoDto;
import com.tiendaya.models.Producto;
import com.tiendaya.interfaces.IProductoService;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping("/v1/productos")
public class ProductoController {

    private final IProductoService productoService;

    public ProductoController(IProductoService productoService) {
        this.productoService = productoService;
    }

    @GetMapping
    public ResponseEntity<List<Producto>> getProductos() {
        List<Producto> productos = productoService.getProductos();

        return ResponseEntity.ok(productos);
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getProducto(@PathVariable Integer id) {
        Optional<Producto> producto = productoService.getProducto(id);

        if (producto.isEmpty()) {
            return ResponseEntity
                    .status(404)
                    .body(new MessageDto("El producto con id " + id + " no existe"));
        }

        return ResponseEntity.ok(producto.get());
    }

    @PostMapping
    public ResponseEntity<Producto> createProducto(@Valid @RequestBody CreateProductoDto dto) {
        Producto producto = productoService.createProducto(dto);

        return ResponseEntity.status(201).body(producto);
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> updateProducto(
            @PathVariable Integer id,
            @RequestBody UpdateProductoDto dto
    ) {
        Optional<Producto> producto = productoService.updateProducto(id, dto);

        if (producto.isEmpty()) {
            return ResponseEntity
                    .status(404)
                    .body(new MessageDto("El producto con id " + id + " no existe"));
        }

        return ResponseEntity.ok(producto.get());
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteProducto(@PathVariable Integer id) {
        Optional<Producto> producto = productoService.deleteProducto(id);

        if (producto.isEmpty()) {
            return ResponseEntity
                    .status(404)
                    .body(new MessageDto("El producto con id " + id + " no existe"));
        }

        return ResponseEntity.ok(new MessageDto("Producto desactivado correctamente"));
    }
}

File: backend/src/controllers/producto.controller.ts
import type { Request, Response } from "express";
import { ProductoService } from "../services/producto.service.js";

export class ProductoController {
  private productoService: ProductoService;

  constructor() {
    this.productoService = new ProductoService();
  }

  getProductos = async (_req: Request, res: Response) => {
    try {
      const productos = await this.productoService.getProductos();

      return res.status(200).json(productos);
    } catch (error) {
      return res.status(500).json({
        mensaje: "Error al obtener productos",
        error
      });
    }
  };

  getProducto = async (req: Request, res: Response) => {
    try {
      const id = Number(req.params.id);

      const producto = await this.productoService.getProducto(id);

      if (!producto) {
        return res.status(404).json({
          mensaje: `El producto con id ${id} no existe`
        });
      }

      return res.status(200).json(producto);
    } catch (error) {
      return res.status(500).json({
        mensaje: "Error al obtener producto",
        error
      });
    }
  };

  createProducto = async (req: Request, res: Response) => {
    try {
      const producto = await this.productoService.createProducto(req.body);

      return res.status(201).json(producto);
    } catch (error) {
      return res.status(500).json({
        mensaje: "Error al crear producto",
        error
      });
    }
  };

  updateProducto = async (req: Request, res: Response) => {
    try {
      const id = Number(req.params.id);

      const producto = await this.productoService.updateProducto(id, req.body);

      if (!producto) {
        return res.status(404).json({
          mensaje: `El producto con id ${id} no existe`
        });
      }

      return res.status(200).json(producto);
    } catch (error) {
      return res.status(500).json({
        mensaje: "Error al actualizar producto",
        error
      });
    }
  };

  deleteProducto = async (req: Request, res: Response) => {
    try {
      const id = Number(req.params.id);

      const producto = await this.productoService.deleteProducto(id);

      if (!producto) {
        return res.status(404).json({
          mensaje: `El producto con id ${id} no existe`
        });
      }

      return res.status(200).json({
        mensaje: "Producto desactivado correctamente",
        producto
      });
    } catch (error) {
      return res.status(500).json({
        mensaje: "Error al eliminar producto",
        error
      });
    }
  };
}